{J,CS}SResourceLocator: account for symlinks in app path

kyrofa opened the pull request:
Currently, if the app path includes a symlink, the calculated webDir will be incorrect when generating CSS and URLs will be pointing to the wrong place, breaking CSS.

This PR fixes #6028 by using realpath when retrieving app path, which makes these issues go away.

kyrofa committed:
CSSResourceLocator: account for symlinks in app path

Currently, if the app path includes a symlink, the calculated webDir
will be incorrect when generating CSS and URLs will be pointing to the
wrong place, breaking CSS.

Use realpath when retrieving app path, and these issues go away.

Fix #6028

Signed-off-by: Kyle Fazzari <kyrofa@ubuntu.com>

kyrofa commented:
I'd love to cover this in a test if anyone can point me in the right direction.

MorrisJobke commented:
> I'd love to cover this in a test if anyone can point me in the right direction.

cc @skjnldsv @juliushaertl

MorrisJobke commented:
I guess we need the same for the JSCombiner?

skjnldsv commented:
Is this related to #5289?

juliushaertl commented:
> I'd love to cover this in a test if anyone can point me in the right direction.

I guess you could write a unit test for that with creating a symlink using http://php.net/manual/en/function.symlink.php. It seems that the CSSResourceLocator class is not unit-tested at all: https://github.com/nextcloud/server/tree/master/tests/lib/Template

kyrofa commented:
> Is this related to #5289?

I'm not sure. It's certainly possible.

kyrofa commented:
@MorrisJobke I'm assuming you're talking about the JSResourceLocator? Assuming so, I've pushed that change here as well.

@juliushaertl I've also pushed tests for both JSResourceLocator as well as CSSResourceLocator that tests the functionality introduced here.

kyrofa committed:
JSResourceLocator: account for symlinks in app path

Signed-off-by: Kyle Fazzari <kyrofa@ubuntu.com>

MorrisJobke commented:
> @juliushaertl I've also pushed tests for both JSResourceLocator as well as CSSResourceLocator that cover the functionality introduced here. I assume they're run automatically in CI, right? Or do I need to alter another file somewhere?

They run automatically (if they are located in a tests/ folder), but cause a Fatal error: 

```
PHP Fatal error:  Cannot redeclare Test\Template\randomString() (previously declared in /drone/src/github.com/nextcloud/server/tests/lib/Template/CSSResourceLocatorTest.php:47) in /drone/src/github.com/nextcloud/server/tests/lib/Template/JSResourceLocatorTest.php on line 46
```

kyrofa committed:
tests: change helper scope

This seems to be the only way to have the same helpers used between
tests in a manner that works for both standalone phpunit and
autotest.sh.

Signed-off-by: Kyle Fazzari <kyrofa@ubuntu.com>

kyrofa commented:
> They run automatically (if they are located in a tests/ folder), but cause a Fatal error:

Thanks @MorrisJobke, all fixed up and ready for review.

juliushaertl commented:
@kyrofa Do you have some steps to reproduce the issue/test the pr? Symlinking the apps folder doesn't cause the issue on my setup.

kyrofa commented:
Sure thing @juliushaertl. Note that while the fix is done in the ResourceLocators, the bug doesn't bite until CSS is generated (which is then cached). The problem isn't quite as simple as "app paths with symlinks don't work", the app path in question actually needs to be within a webroot that is a symlink, so let's use the default apps as an example.

1. Get a completely clean, not-installed Nextcloud
2. The webroot needs to be a symlink, so instead of `/var/www/nextcloud`, put Nextcloud in `/var/www/actual-nextcloud`.
3. Now create `/var/www/nextcloud` as a symlink pointing to `/var/www/actual-nextcloud` (webserver should be pointing to the symlink).
4. Before installing, add an `apps_paths` item pointing to the default apps, but use the symlink path for it, not the real one, e.g.:
```
<?php
$CONFIG = array (
  'instanceid' => 'oc4ugkbj34v5',
  'apps_paths' => array(
        array(
                # Note the use of nextcloud instead of actual-nextcloud
                'path'=> '/var/www/nextcloud/apps',
                'url' => '/apps',
                'writable' => true,
        ),
  ),
);
```

Now you've set things up such that:

1. The CSS has not yet been generated since Nextcloud is not installed
2. The webroot, which is a symlink, will be resolved into `/var/www/actual-nextcloud`
3. When the CSS is generated, that webroot will be used to rewrite URLs for the apps, but the app path will still be `/var/www/nextcloud/apps/foo` and the rewrite will go sideways, creating invalid URLs to assets, thus bad CSS.

Go ahead and install Nextcloud now, and you'll see things like this:

![image](https://user-images.githubusercontent.com/946674/32634505-decc2138-c560-11e7-80ad-7877f9a60bd1.png)

juliushaertl commented on lib/private/Template/JSResourceLocator.php:
```diff
@@ -75,9 +75,14 @@ public function doFind($script) {
+		// Account for the possibility of having symlinks in app path. Doing
+		// this in a separate variable, because an empty argument to realpath
+		// gets turned into cwd, which makes it hard to see if app_path got set.
+		$real_app_path = realpath($app_path);
```
Shouldn't we also do this after the check if the app_path is empty like it is done in in the CSSResourceLocator? Otherwise $real_app_path still might be the working directory for empty app paths.

kyrofa replied:
Yeah the diff is hiding that from you, expand is a little below this. It's done [here](https://github.com/nextcloud/server/pull/7061/files#diff-e75d752a35a50ba300698850b35f80efR89).

kyrofa replied:
Although this will cause weird behavior with using the cwd for l10n, huh. I refactored it just a tad to account for that.

juliushaertl replied:
Can't you just move the check got app_path and app_url in front of the realpath call? That way we don't append any js sources at cwd.

kyrofa replied:
Not without changing the `l10n` behavior. I don't claim to be an expert in the code, of course, but right now if a `l10n` file doesn't exist, the error is ignored. If we move the check above that logic, we start raising an error instead.

juliushaertl replied:
Ah, ok, makes sense to keep it like this i guess. 👍

juliusknorr reviewed (changes requested):
Thanks @kyrofa I could reproduce it and your fix works great for that.

kyrofa committed:
JSResourceLocator: re-use $app_path

Signed-off-by: Kyle Fazzari <kyrofa@ubuntu.com>

MorrisJobke commented:
The failed unit test is unrelated.

kyrofa commented:
Can we get this back into stable12 as well? It's the last remaining issue for the snap.

MorrisJobke commented:
> Can we get this back into stable12 as well? It's the last remaining issue for the snap.

Could you open a backport to the stable12 branch as PR?

kyrofa commented:
> Could you open a backport to the stable12 branch as PR?

Sure thing! #7170
